superset/utils/pandas_postprocessing.py

```diff
@@ -0,0 +1,210 @@
+
+    df = df.pivot_table(
+        values=aggfunc.keys(),
+        index=index,
+        columns=columns,
+        aggfunc=aggfunc,
+        fill_value=metric_fill_value,
+        dropna=drop_missing_columns,
+        margins=marginal_distributions,
+        margins_name=marginal_distribution_name,
+    )
+
+    if combine_value_with_metric:
+        df = df.stack(0).unstack()
+
+    return df
+
+
+def aggregate(
+    df: DataFrame, groupby: List[str], aggregates: Dict[str, Dict[str, Any]]
+) -> DataFrame:
+    """
+    Apply aggregations to a DataFrame.
+
+    :param df: Object to aggregate.
+    :param groupby: columns to aggregate
+    :param aggregates: A mapping from metric column to the function used to
+           aggregate values.
+    :return: Aggregated DataFrame
+    """
+    aggregates = aggregates or {}
+    aggregate_funcs = _get_aggregate_funcs(aggregates)
+    return df.groupby(by=groupby).agg(**aggregate_funcs)
+
+
+def sort(
+    df: DataFrame,
+    by: List[str],
+    ascending: Optional[Union[bool, Dict[str, bool]]] = True,
+) -> DataFrame:
+    """
+    Sort a DataFrame.
+
+    df: DataFrame to sort.
+    by: columns by by which to sort `df`.
+    ascending: Sort order. Defaults to True. If bool, applies same sort order to all
+               columns. If dict, applies sorting per column, defaulting to True if
+               missing.
+    """
+    if isinstance(ascending, dict):
+        ascending = [ascending.get(col, True) for col in by]
+
+    return df.sort_values(by=by, ascending=ascending)
+
+
+def rolling(
+    df: DataFrame,
+    rolling_type: str,
+    center: bool = False,
+    win_type: Optional[str] = None,
+    window: Optional[int] = None,
+    min_periods: Optional[int] = None,
+) -> DataFrame:
+    """
+    Apply a rolling window on the dataset. See the Pandas docs for further details:
+    https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rolling.html
+
+    :param df: DataFrame on which the rolling period will be based.
+    :param rolling_type: Type of rolling window. Any numpy function will work.
+    :param center: Should the label be at the center of the window.
+    :param win_type: Type of window function.
+    :param window: Size of the window.
+    :param min_periods:
+    :return:
+    """
+    if rolling_type == "cumsum":
+        df = df.cumsum()
+    else:
+        kwargs = {}
+        if window is not None:
+            kwargs["window"] = window
+        if min_periods is not None:
+            kwargs["min_periods"] = min_periods
+        if center is not None:
+            kwargs["center"] = center
+        if win_type is not None:
+            kwargs["win_type"] = win_type
+
+        df_rolling = df.rolling(**kwargs)
+        if not hasattr(df_rolling, rolling_type):
+            raise SupersetException(
+                _("Unsupported rolling_type: %(type)s", type=rolling_type)
+            )
+        df = getattr(df_rolling, rolling_type)()
```

Comment:
It would be more useful if

1. Rolling results are added as a new column to the original df
2. Users can select which columns to roll
3. There is the ability to compute rolling diffs

How about an API like this:

```python
[
    {
        "operation": "rolling",
        "options": {
            "columns": ["abc", "cde"],
            # 7 day moving average
            "output_columns": ["abc MA7", "cde MA7"],
            "rolling_type": "mean",
            "win_type": None,
            "window": 7,
            "min_periods": 7,
        },
    },
    {
        "operation": "rolling",
        "options": {
            "columns": ["abc", "cde"],
            # MoM growth computed with 30 MA
            "output_columns": ["abc MoM", "cde MoM"],
            "rolling_type": "mean",
            "win_type": None,
            "window": 30,
            "min_periods": 30,
            # apply `df.diff(...)` if `diff_periods is specified`
            "diff_periods": 30,
        },
    },
    {
        "operation": "rolling",
        "options": {
            "columns": ["abc", "cde"],
            # YoY growth in percentages
            "output_columns": ["abc YoY (%)", "cde YoY (%)"],
            "rolling_type": "mean",
            "win_type": None,
            "window": 365,
            "min_periods": 365,
            # apply `df.pct_change(...)`
            "pct_change_periods": 365,
        },
    },
    # `diff_periods` and `pct_change_periods` are shortcuts
    # to applying additional `diff` and `pct_change` operators
    # on the output columns.
    {
        "operation": "diff",
        "options": {
            "columns": ["abc", "cde"],
            "output_columns": ["abc MoM", "cde MoM"],
            "periods": 7,
        },
    },
]
```

Reply:
Thanks @ktmud , these are great suggestions. I think there's room for adding lots of functionality here, but I would prefer to keep the initial PR as compact as possible. How about I make sure all operations support adding new columns (where applicable), but we leave the new diff operation to a later PR?

Reply:
Sgtm

Reply:
This is great work, btw! Really looking forward to use these API in the charts.